pandas/core/arrays/categorical.py

```diff
@@ -2423,11 +2424,26 @@ def _get_codes_for_values(values, categories):
     utility routine to turn values into codes given the specified categories
     """
     from pandas.core.algorithms import _get_data_algo, _hashtables
-    if is_dtype_equal(values.dtype, categories.dtype):
+    dtype_equal = is_dtype_equal(values.dtype, categories.dtype)
+
+    if is_extension_array_dtype(categories.dtype) and is_object_dtype(values):
+        # Support inferring the correct extension dtype from an array of
```

Comment:
I would make this a function  I think as we are likely to do this in other places as well (or maybe already) do, e.g. in csv parsing with category?

Reply:
this is basically what ``catetorical_array`` (generic construtor for Categoricals should be doing anyhow)

Reply:
This called from the categorical constructor.

Ideally we would have a single place that does all this. Right now it feels scattered over `_sanitize_array`, the Index constructor, and probably a few other places.